### PR TITLE
More restrictive deserialization and add test cases

### PR DESCRIPTION
**Pull request by yarkinwho**
Part of the refactor effort of related leap crypto functions.

### CODE REVIEW

File: src/g.cpp [hunk 1]
```diff
@@ -19,9 +19,9 @@ g1::g1(const g1& e) : x(e.x), y(e.y), z(e.z)
...
 optional<g1> g1::fromJacobianBytesBE(const span<const uint8_t, 144> in, const bool check, const bool raw)
 {
-    optional<fp> x = fp::fromBytesBE(span<const uint8_t, 48>(&in[ 0], &in[ 48]), check, raw);
-    optional<fp> y = fp::fromBytesBE(span<const uint8_t, 48>(&in[48], &in[ 96]), check, raw);
-    optional<fp> z = fp::fromBytesBE(span<const uint8_t, 48>(&in[96], &in[144]), check, raw);
+    optional<fp> x = fp::fromBytesBE(span<const uint8_t, 48>(&in[ 0], &in[ 48]), true, raw);
+    optional<fp> y = fp::fromBytesBE(span<const uint8_t, 48>(&in[48], &in[ 96]), true, raw);
+    optional<fp> z = fp::fromBytesBE(span<const uint8_t, 48>(&in[96], &in[144]), true, raw);
     if(!x.has_value() || !y.has_value() || !z.has_value()) return {};
     g1 p = g1({x.value(), y.value(), z.value()});
```
Comment (greg7mdp) [inline src/g.cpp:26]: That is a non-idiomatic (and verbose) way to use std::optional. I would prefer:
```suggestion
    if(!x || !y || !z) return {};
    g1 p({*x, *y, *z});
```

File: src/g.cpp [hunk 2]
```diff
@@ -699,9 +700,9 @@ g2::g2(const g2& e) : x(e.x), y(e.y), z(e.z)
 
 optional<g2> g2::fromJacobianBytesBE(const span<const uint8_t, 288> in, const bool check, const bool raw)
 {
-    optional<fp2> x = fp2::fromBytesBE(span<const uint8_t, 96>(&in[  0], &in[ 96]), check, raw);
-    optional<fp2> y = fp2::fromBytesBE(span<const uint8_t, 96>(&in[ 96], &in[192]), check, raw);
-    optional<fp2> z = fp2::fromBytesBE(span<const uint8_t, 96>(&in[192], &in[288]), check, raw);
+    optional<fp2> x = fp2::fromBytesBE(span<const uint8_t, 96>(&in[  0], &in[ 96]), true, raw);
+    optional<fp2> y = fp2::fromBytesBE(span<const uint8_t, 96>(&in[ 96], &in[192]), true, raw);
+    optional<fp2> z = fp2::fromBytesBE(span<const uint8_t, 96>(&in[192], &in[288]), true, raw);
```
Comment (greg7mdp) [inline src/g.cpp:705]: Add comment:
`// We decided to always validate the input here. But we reserve the flag.`

File: src/g.cpp [hunk 3]
```diff
@@ -713,9 +714,9 @@ optional<g2> g2::fromJacobianBytesBE(const span<const uint8_t, 288> in, const bo
 
 optional<g2> g2::fromJacobianBytesLE(const span<const uint8_t, 288> in, const bool check, const bool raw)
 {
-    optional<fp2> x = fp2::fromBytesLE(span<const uint8_t, 96>(&in[  0], &in[ 96]), check, raw);
-    optional<fp2> y = fp2::fromBytesLE(span<const uint8_t, 96>(&in[ 96], &in[192]), check, raw);
-    optional<fp2> z = fp2::fromBytesLE(span<const uint8_t, 96>(&in[192], &in[288]), check, raw);
+    optional<fp2> x = fp2::fromBytesLE(span<const uint8_t, 96>(&in[  0], &in[ 96]), true, raw);
+    optional<fp2> y = fp2::fromBytesLE(span<const uint8_t, 96>(&in[ 96], &in[192]), true, raw);
+    optional<fp2> z = fp2::fromBytesLE(span<const uint8_t, 96>(&in[192], &in[288]), true, raw);
```
Comment (greg7mdp) [inline src/g.cpp:719]: Add comment here and two next occurences:
`// We decided to always validate the input here. But we reserve the flag.`

File: src/g.cpp [hunk 4]
```diff
@@ -952,7 +957,7 @@ bool g2::inCorrectSubgroup() const
     t0 = t0.psi();
     t1 = t0.neg();                  // - ψ^2(P)
     t0 = t0.psi();                  // ψ^3(P)
-    t0 = t0.mulScalar(cofactorEFF); // - x ψ^3(P)
+    t0 = t0.scale(cofactorEFF); // - x ψ^3(P)
```
Comment (greg7mdp) [inline src/g.cpp:960]: align comment

File: src/g.cpp [hunk 5]
```diff
@@ -1164,10 +1169,10 @@ g2 g2::frobeniusMap(int64_t power) const
     return r;
 }
 
-// MultiExp calculates multi exponentiation. Given pairs of G2 point and scalar values
+// WeightedSum calculates multi exponentiation. Given pairs of G2 point and scalar values
 // (P_0, e_0), (P_1, e_1), ... (P_n, e_n) calculates r = e_0 * P_0 + e_1 * P_1 + ... + e_n * P_n
 // Length of points and scalars are expected to be equal, otherwise NONE is returned.
-optional<g2> g2::multiExp(const vector<g2>& points, const vector<array<uint64_t, 4>>& scalars, std::function<void()> yield)
+optional<g2> g2::weightedSum(const vector<g2>& points, const vector<array<uint64_t, 4>>& scalars, std::function<void()> yield)
```
Comment (greg7mdp) [inline src/g.cpp:1175]: Make `yield` a `const&`. Same for `g1` version. Perf issue as yield is unnecessarily copied for each invocation.

File: test/unittests.cpp [hunk 6]
```diff
@@ -738,6 +840,35 @@ void TestG2Serialization()
     }
 }
 
+void TestG2SerializationGarbage() {
+    array<uint8_t, 288> buf;
+    buf.fill(0xff);
```
Comment (greg7mdp) [inline test/unittests.cpp:845]: Instead of just testing only with all `0xff`, we could also loop over a large number (10k?) of  randomly generated values?

File: src/arithmetic.cpp [hunk 7]
```diff
@@ -1796,7 +1507,7 @@ extern "C" blsmul_func_t __attribute__((no_sanitize_address)) resolve_blsmul() {
     return __mul;
```
Comment (greg7mdp) [inline src/arithmetic.cpp:1507]: `__mul` not found. Renamed to `__multiply`?

**Review by greg7mdp — CHANGES_REQUESTED** [review]

File: include/bls12-381/fp.hpp [hunk 8]
```diff
@@ -36,6 +36,19 @@ class fp
...
+    
+    fp add(const fp& e) const;
+    void addAssign(const fp& e);
+    fp dbl() const;
+    void doubleAssign();
+    fp subtract(const fp& e) const;
+    void subtractAssign(const fp& e);
+    fp negate() const;
+    fp multiply(const fp& e) const;
+    void multiplyAssign(const fp& e);
```
Comment (greg7mdp) [inline include/bls12-381/fp.hpp:49]: What about using the more intuitive versions:

```
    fp operator+(const fp& e) const;
    void operator+=(const fp& e);
    fp dbl() const;
    void doubleAssign();
    fp operator-(const fp& e) const;
    void operator-=(const fp& e);
    fp operator-() const;
    fp operator*(const fp& e) const;
```

File: src/fp.cpp [hunk 9]
```diff
@@ -19,6 +19,7 @@ fp::fp(const fp& e) : d{e.d[0], e.d[1], e.d[2], e.d[3], e.d[4], e.d[5]}
 
 optional<fp> fp::fromBytesBE(const span<const uint8_t, 48> in, const bool check, const bool raw)
 {
+    // We decided to always validate the input here. But we reserve the flag.
     fp e = fp(scalar::fromBytesBE<6>(in));
```
[the remark this answers is not in the document]
Comment (yarkinwho) [reply src/fp.cpp:23]: The way we enforcing checks has been changed. The function no longer has this issue.

File: src/fp.cpp [hunk 10]
```diff
@@ -27,6 +28,7 @@ optional<fp> fp::fromBytesBE(const span<const uint8_t, 48> in, const bool check,
 
 optional<fp> fp::fromBytesLE(const span<const uint8_t, 48> in, const bool check, const bool raw)
 {
+    // We decided to always validate the input here. But we reserve the flag.
     fp e = fp(scalar::fromBytesLE<6>(in));
```
[the remark this answers is not in the document]
Comment (yarkinwho) [reply src/fp.cpp:32]: The way we enforcing checks has been changed. The function no longer has this issue.

File: src/g.cpp (hunk 4 above)
Comment (yarkinwho) [reply src/g.cpp]: done

File: src/g.cpp (hunk 3 above)
Comment (yarkinwho) [reply src/g.cpp:722]: done

File: src/g.cpp (hunk 2 above)
Comment (yarkinwho) [reply src/g.cpp:707]: done

File: src/arithmetic.cpp (hunk 7 above)
Comment (yarkinwho) [reply src/arithmetic.cpp]: fixed

File: src/g.cpp (hunk 1 above)
Comment (yarkinwho) [reply src/g.cpp]: fixed all similar issues.

File: include/bls12-381/fp.hpp (hunk 8 above)
Comment (yarkinwho) [reply include/bls12-381/fp.hpp:49]: These functions are here to make them align with fp2, fp6 etc.

The will not really be used in our code for performance reason.

If we want operators for fp, then maybe we also want operators for fp2 fp6 and maybe fp12. Then the problem becomes is this really necessary...

Comment (greg7mdp) [reply include/bls12-381/fp.hpp:49]: Personally, I prefer using the intuitive operators, rather than invent new function names whose usage is less straightforward imo.
But I am not requesting this to be changed.

File: src/g.cpp (hunk 5 above)
Comment (yarkinwho) [reply src/g.cpp]: fixed

File: test/unittests.cpp (hunk 6 above)
Comment (yarkinwho) [reply test/unittests.cpp:845]: This test is meant to test corner cases. There's another test testing random inputs.

File: src/g.cpp [hunk 11]
```diff
@@ -460,39 +478,45 @@ optional<g1> g1::multiExp(std::span<const g1> points, std::span<const array<uint
     {
         for(uint64_t i = 0; i < bucketSize; i++)
         {
+            if (i & 255 == 0) {
```
Comment (greg7mdp) [inline src/g.cpp:481]: operator `==` has higher precedence than bitwise and `&`, so this test is equivalent to i & (255 == 0) which is probably not what was intended.
So either parenthesize `(i & 255)` or use `i % 256` and let the compiler optimize :-).

Comment (greg7mdp) [inline src/g.cpp:481]: We should probably check that yield is not null:
```suggestion
            if (i % 256 == 0 && yield) {
```

**Review by greg7mdp — CHANGES_REQUESTED** [review]

File: src/g.cpp (hunk 11 above)
Comment (yarkinwho) [reply src/g.cpp]: change to if (yield && ((i & 255) == 0))

Comment (yarkinwho) [reply src/g.cpp]: change to if (yield && ((i & 255) == 0))

File: test/unittests.cpp [hunk 12]
```diff
@@ -607,40 +821,54 @@ void TestG1MultiExpExpected()
...
-    }
+    const auto doTest = [](int64_t n) {
+        g1 one = g1::one();
+        vector<array<uint64_t, 4>> scalars;
+        vector<g1> bases;
+        // scalars: [s0,s1 ... s(n-1)]
+        // bases: [P0,P1,..P(n-1)] = [s(n-1)*G, s(n-2)*G ... s0*G]
+        for(int64_t i = 0; i < n; i++)
+        {
+            scalars.insert(scalars.begin(), array<uint64_t, 4>{static_cast<uint64_t>(rand()%100000), 0, 0, 0});
```
Comment (greg7mdp) [inline test/unittests.cpp:842]: why `insert` instead of `emplace_back()`?

File: test/unittests.cpp [hunk 13]
```diff
@@ -607,40 +821,54 @@ void TestG1MultiExpExpected()
...
+            bases.push_back(g1::zero());
+            bases[i] = one.scale(array<uint64_t, 4>{static_cast<uint64_t>(rand()%100000), 0, 0, 0});
```
Comment (greg7mdp) [inline test/unittests.cpp:844]:
```suggestion
            bases.push_back(one.scale(array<uint64_t, 4>{static_cast<uint64_t>(rand()%100000), 0, 0, 0}));
```

File: test/unittests.cpp (hunk 12 above)
Comment (yarkinwho) [reply test/unittests.cpp:881]: just copied from existing test code... let me double check what it's doing...

File: test/unittests.cpp (hunk 13 above)
Comment (yarkinwho) [reply test/unittests.cpp]: now it's 
for(int64_t i = 0; i < n; i++)
        {
            scalars.push_back(random_scalar());
            bases.push_back(random_g1());
        }

File: include/bls12-381/fp.hpp [hunk 14]
```diff
@@ -19,10 +19,10 @@ class fp
     std::array<uint64_t, 6> d;
 
     fp();
-    fp(const std::array<uint64_t, 6>& d);
+    explicit fp(const std::array<uint64_t, 6>& d);
```
Comment (greg7mdp) [inline include/bls12-381/fp.hpp:22]: Whay make it explicit?

File: include/bls12-381/fp.hpp [hunk 15]
```diff
@@ -74,10 +87,10 @@ class fp2
     fp c1;
 
     fp2();
-    fp2(const std::array<fp, 2>& e2);
+    explicit fp2(const std::array<fp, 2>& e2);
```
Comment (greg7mdp) [inline include/bls12-381/fp.hpp:90]: Why explicit?

File: include/bls12-381/fp.hpp (hunk 14 above)
Comment (yarkinwho) [reply include/bls12-381/fp.hpp:22]: We only do the checks and Montgomery transform in deserialization methods. We and to reduce the probability of fp being constructed from raw array. explicit will help this slightly.

File: include/bls12-381/fp.hpp (hunk 15 above)
Comment (yarkinwho) [reply include/bls12-381/fp.hpp:90]: We only do the checks and Montgomery transform in deserialization methods. We and to reduce the probability of fp being constructed from raw array. explicit will help this slightly.

File: include/bls12-381/fp.hpp (hunk 14 above)
Comment (greg7mdp) [reply include/bls12-381/fp.hpp:22]: Hum, I don't think it is very likely that you would pass a `std::array<uint64_t, 6>` without expecting to create a `fp`, but ok.

**Review by greg7mdp — APPROVED** [review]

File: src/g.cpp [hunk 16]
```diff
@@ -1158,25 +1206,22 @@ g2 g2::frobeniusMap(int64_t power) const
...
-        r.x = r.x.mul(frb0);
-        r.y = r.y.mul(frb1);
+        r.x = r.x.multiply(frb0);
+        r.y = r.y.multiply(frb1);
     }
     return r;
 }
 
-// MultiExp calculates multi exponentiation. Given pairs of G2 point and scalar values
+// WeightedSum calculates multi exponentiation. Given pairs of G2 point and scalar values
```
Comment (arhag) [inline src/g.cpp:1215]: Remove `WeightedSum calculates multi exponentiation.` from comment.

File: src/g.cpp [hunk 17]
```diff
@@ -1158,25 +1206,22 @@ g2 g2::frobeniusMap(int64_t power) const
...
 // (P_0, e_0), (P_1, e_1), ... (P_n, e_n) calculates r = e_0 * P_0 + e_1 * P_1 + ... + e_n * P_n
 // Length of points and scalars are expected to be equal, otherwise NONE is returned.
```
Comment (arhag) [inline src/g.cpp:1217]: Replace

```
Length of points and scalars are expected to be equal, otherwise NONE is returned.
```

with

```
If length of points and scalars are not the same, 
then missing points will be treated as the zero point 
and missing scalars will be treated as the zero scalar.
```

File: test/unittests.cpp [hunk 18]
```diff
@@ -204,6 +204,297 @@ void TestFieldElementEquality()
...
+        fp as = a.square();
+        fp asqrt;
+        if (as.sqrt(asqrt)) {
+            if(!as.equal(asqrt.square()))
+            {
+                throw invalid_argument("sqrt(fp).square != fp");
+            }
+            if(!a.equal(asqrt) && !a.equal(asqrt.negate()))
+            {
+                throw invalid_argument("fp2!= sqrt(fp.square)");
```
Comment (arhag) [inline test/unittests.cpp:403]: Error message should say `fp != sqrt(fp.square)`.

File: test/unittests.cpp [hunk 19]
```diff
@@ -607,40 +928,53 @@ void TestG1MultiExpExpected()
...
+        g1 expected, tmp;
+        for(int64_t i = 0; i < n; i++)
+        {
+            tmp = bases[i].scale(scalars[i]);
+            expected = expected.add(tmp);
+        }
+        g1 result = g1::weightedSum(bases, scalars);
+        if(!expected.equal(result))
+        {
+            throw invalid_argument("bad multi-exponentiation");
```
Comment (arhag) [inline test/unittests.cpp:962]: Error message should say `"bad g1 weighted sum"`.

Also, the comments in this function are misleading.

File: test/unittests.cpp [hunk 20]
```diff
@@ -891,40 +1254,53 @@ void TestG2MultiExpExpected()
...
+        g2 expected, tmp;
+        for(int64_t i = 0; i < n; i++)
+        {
+            tmp = bases[i].scale(scalars[i]);
+            expected = expected.add(tmp);
+        }
+        g2 result = g2::weightedSum(bases, scalars);
+        if(!expected.equal(result))
+        {
+            throw invalid_argument("bad multi-exponentiation");
```
Comment (arhag) [inline test/unittests.cpp:1288]: Error message should say `"bad g2 weighted sum"`.

Also, the comments in this function are misleading.

File: src/fp.cpp [hunk 21]
```diff
@@ -445,133 +514,101 @@ bool fp2::sign() const
...
 {
     fp t[3];
     _ladd(&t[0], &c0, &c1);
-    _sub(&t[1], &c0, &c1);
+    _subtract(&t[1], &c0, &c1);
     _ldouble(&t[2], &c0);
-    _mul(&c0, &t[0], &t[1]);
-    _mul(&c1, &t[2], &c1);
+    _multiply(&c0, &t[0], &t[1]);
+    _multiply(&c1, &t[2], &c1);
```
Comment (arhag) [inline src/fp.cpp:604]: Have the above two `_multiply` calls been verified to be safe?

`t[0]` and `[2]` could both be larger than the prime modulus.

File: src/fp.cpp [hunk 22]
```diff
@@ -210,10 +279,10 @@ fp fp::inverse() const
 
     if(r.cmp(MODULUS) >= 0 || z > 0)
     {
-        _lsubAssign(&r, &MODULUS);
+        _lsubtract(&r, &r, &MODULUS);
     }
     u = MODULUS;
-    _lsubAssign(&u, &r);
+    _lsubtract(&u, &u, &r);
```
Comment (arhag) [inline src/fp.cpp:285]: At this point, I believe it is possible for `u` to be greater than or equal to the prime modulus. For example, consider the case where `r` was zero at this point (not sure if it is possible to reach this point with a `r` of zero). Or consider the case where `r` is greater than the prime modulus (again not sure if that is possible).

If this is true, then this is concerning given that it will be used in a call to `_double` below. However, it may be safe to call `_double` with a field element greater than or equal to the prime modulus and still get a mathematically sensible output. But has that been verified to be the case?

File: test/unittests.cpp (hunk 20 above)
Comment (yarkinwho) [reply test/unittests.cpp]: fixed

File: test/unittests.cpp (hunk 19 above)
Comment (yarkinwho) [reply test/unittests.cpp]: fixed

File: test/unittests.cpp (hunk 18 above)
Comment (yarkinwho) [reply test/unittests.cpp]: fixed

File: src/g.cpp (hunk 17 above)
Comment (yarkinwho) [reply src/g.cpp]: fixed

File: src/g.cpp (hunk 16 above)
Comment (yarkinwho) [reply src/g.cpp]: fixed

File: src/g.cpp [hunk 23]
```diff
@@ -436,19 +457,16 @@ g1 g1::glvEndomorphism() const
     return t;
 }
 
-// MultiExp calculates multi exponentiation. Given pairs of G1 point and scalar values
+// WeightedSum calculates multi exponentiation. Given pairs of G1 point and scalar values
 // (P_0, e_0), (P_1, e_1), ... (P_n, e_n) calculates r = e_0 * P_0 + e_1 * P_1 + ... + e_n * P_n
 // Length of points and scalars are expected to be equal, otherwise NONE is returned.
```
[the remark this answers is not in the document]
Comment (yarkinwho) [reply src/g.cpp]: fixed

File: src/g.cpp [hunk 24]
```diff
@@ -436,19 +457,16 @@ g1 g1::glvEndomorphism() const
     return t;
 }
 
-// MultiExp calculates multi exponentiation. Given pairs of G1 point and scalar values
+// WeightedSum calculates multi exponentiation. Given pairs of G1 point and scalar values
```
[the remark this answers is not in the document]
Comment (yarkinwho) [reply src/g.cpp]: fixed

File: src/fp.cpp (hunk 22 above)
Comment (yarkinwho) [reply src/fp.cpp:291]: Checked and optimized. See the comments in the function.

**Comment by yarkinwho** [comment]
One last issue I noticed:
@greg7mdp @arhag 

I noticed default comparisons are present for fp, fp2, g1 and g2

I think it would be ambiguous to define < and > for fp2/g1/g2 and the default = (lexicographical comparison) would be wrong for g1/g2. = should be defined by equal() instead of compare xyz as fp/fp2.

**Review by arhag — APPROVED** [review]
